### PR TITLE
Video reference scripts

### PR DESCRIPTION
**Pull request by fmassa**
This PR adds training and evaluation scripts for video models.

It also adds a few extra helper functions, which should ideally be integrated in PyTorch / Torchvision instead of being part of the reference scripts. For now they are added here to avoid having to worry about backwards-compatibility.

Some parts of the main training script needs cleanup, specially the part handling caching of the dataset.
I'm sending this PR now for early feedback.

Note that the first commit is only copying as is the training scripts from image classification, and do not need to be reviewed.

cc @bjuncek

### CODE REVIEW

File: references/video_classification/train.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+            # FIXME need to take into account that the datasets
+            # could have been padded in distributed setup
+            batch_size = image.shape[0]
+            metric_logger.update(loss=loss.item())
+            metric_logger.meters['acc1'].update(acc1.item(), n=batch_size)
+            metric_logger.meters['acc5'].update(acc5.item(), n=batch_size)
+    # gather the stats from all processes
+    metric_logger.synchronize_between_processes()
+
+    print(' * Acc@1 {top1.global_avg:.3f} Acc@5 {top5.global_avg:.3f}'
```
Comment (bjuncek) [inline references/video_classification/train.py]: Could we just  note that this is clip-level accuracy -  maybe change the printing to Clip@1 to avoid confusion?

File: references/video_classification/train.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+    device = torch.device(args.device)
+
+    torch.backends.cudnn.benchmark = True
+
+    # Data loading code
+    print("Loading data")
+    traindir = os.path.join(args.data_path, 'train_avi-480p')
+    valdir = os.path.join(args.data_path, 'val_avi-480p')
+    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
```
Comment (bjuncek) [inline references/video_classification/train.py]: Did you compute these numbers? They don't seem to match with the VMZ ones.

File: references/video_classification/train.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+                                     std=[0.229, 0.224, 0.225])
+
+    print("Loading training data")
+    st = time.time()
+    cache_path = _get_cache_path(traindir)
+    dataset = torch.load("/private/home/fmassa/github/vision/kinetics_train.pth")
```
Comment (bjuncek) [inline references/video_classification/train.py]: Probably Want to remove hardcoded path?

Comment (bjuncek) [reply references/video_classification/train.py]: Also, do you literally save the dataset object? (that's what I've been doing but I'm not 100% sure if it's ok to do).

File: references/video_classification/train.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                normalize,
+            ]))
+        if args.cache_dataset:
+            print("Saving dataset_test to {}".format(cache_path))
+            utils.mkdir(os.path.dirname(cache_path))
+            utils.save_on_master((dataset_test, valdir), cache_path)
+    """
+    print("Creating data loaders")
+    train_sampler = torchvision.datasets.video_utils.RandomClipSampler(dataset.video_clips, args.clips_per_video)
+    test_sampler = SequentialClipSampler(dataset_test.video_clips, args.clips_per_video)
```
Comment (bjuncek) [inline references/video_classification/train.py]: Do we plan to add this and the uniform sampler to the `video_utils`?

File: references/video_classification/train.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+    # per iteration, not per epoch
+    warmup_epochs = 10
+    warmup_iters = warmup_epochs * len(data_loader)
+    lr_milestones = [len(data_loader) * m for m in args.lr_milestones]
+    lr_scheduler = WarmupMultiStepLR(
+        optimizer, milestones=lr_milestones, gamma=args.lr_gamma,
+        warmup_iters=warmup_iters, warmup_factor=1e-5)
+
+    model_without_ddp = model
```
Comment (bjuncek) [inline references/video_classification/train.py]: Why two models?

File: torchvision/datasets/kinetics.py [hunk 6]
```diff
@@ -23,4 +24,8 @@ def __getitem__(self, idx):
         video, audio, info, video_idx = self.video_clips.get_clip(idx)
         label = self.samples[video_idx][1]
 
-        return video, audio, label
+        if self.transform is not None:
+            video = self.transform(video)
+
+        # return video, audio, label
+        return video, label
```
Comment (bjuncek) [inline torchvision/datasets/kinetics.py]: If you are gathering the results, wouldn't you need an index for validation as well?

File: torchvision/io/video.py [hunk 7]
```diff
@@ -95,7 +95,7 @@ def _read_from_stream(container, start_offset, end_offset, stream, stream_name):
         # TODO check if stream needs to always be the video stream here or not
         container.seek(seek_offset, any_frame=False, backward=True, stream=stream)
     except av.AVError:
-        print("Corrupted file?", container.name)
+        # print("Corrupted file?", container.name)
```
Comment (bjuncek) [inline torchvision/io/video.py]: We should probably have some sort of "verbose" flag for this?

File: references/video_classification/train.py (hunk 2 above)
Comment (fmassa) [reply references/video_classification/train.py]: Good catch. This is currently not used, as I overwrite it with the VMZ ones in https://github.com/pytorch/vision/pull/1180/files#diff-0eb452652bb24a2750ae67b48421239bR126

File: references/video_classification/train.py (hunk 3 above)
Comment (fmassa) [reply references/video_classification/train.py]: this is ok to do in general, but it also serializes the transforms.
My initial thinking was to only serialize `VideoClips`, but then I need to let all datasets be constructed by passing a `VideoClips` object, which I'm not sure I want to do.

Thoughts?

File: references/video_classification/train.py (hunk 4 above)
Comment (fmassa) [reply references/video_classification/train.py]: Yes, I didn't add it there because I'm not sure I like the place where the current sampler is. I'm actually thinking about moving the other sampler away from there, and inside `references` for now, while we decide where to put it

File: references/video_classification/train.py (hunk 5 above)
Comment (fmassa) [reply references/video_classification/train.py]: This is just a pointer to the other model without DDP, and is useful for loading / saving the checkpoint without having to worry if the model / checkpoint has DDP or not.

File: torchvision/datasets/kinetics.py (hunk 6 above)
Comment (fmassa) [reply torchvision/datasets/kinetics.py]: Yes, but I'm currently making an approximation during evaluation, where I can have some images / clips which are accumulated multiple times.

I'd rather not return the `index` in the dataset itself, as this is a kind of a hack that I think shouldn't live in torchvision.

If we definitely want to have a more accurate way of handling distributed evaluation, then I think I'll add an extra dataset in `references` that just wraps the `idx` as well, doing something like

```python
class Kinetics2(Kinetics):
    def __getitem__(self, idx):
        video, label = super(Kinetics2, self).__getitem__(idx)
        return video, label, idx
```

File: torchvision/io/video.py (hunk 7 above)
Comment (fmassa) [reply torchvision/io/video.py]: yes, but we currently do not have any type of logging in torchvision. Printing to screen is definitely not something we want to do. And adding a logging option will take some more work and thought and I'd rather not rush things.

@soumith thoughts?

File: references/video_classification/sampler.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        return iter(indices)
+
+    def __len__(self):
+        return self.num_samples
+
+    def set_epoch(self, epoch):
+        self.epoch = epoch
+
+
+class SequentialClipSampler(torch.utils.data.Sampler):
```
Comment (bjuncek) [inline references/video_classification/sampler.py]: Might want to rename to UniformClipSampler, for clarity

File: references/video_classification/train.py (hunk 3 above)
Comment (bjuncek) [reply references/video_classification/train.py]: I agree that passing VideoClips is a pain.
Here is my experimentation workaround:


```
class Ucf101(UCF101):

    def __init__(self, root, annotation_path, frames_per_clip, step_between_clips=1,
                 fold=1, train=True, spatial_transform=None, audio_transform=None):

        super().__init__(root, frames_per_clip, step_between_clips)
        self.audio_transform = audio_transform
        self.spatial_transform = spatial_transform

    def __getitem__(self, idx):
        video, audio, info, video_idx = self.video_clips.get_clip(idx)
        label = self.samples[video_idx][1]

        if self.audio_transform:
            audio = self.audio_transform(audio)
        if self.spatial_transform:
            video = self.spatial_transform(video)

        return video, audio, label, video_idx

    def set_spatial_transform(self, transform):
        self.spatial_transform = transform
    
    def set_audio_transform(self, transform):
        self.audio_transform = transform
```

Not pretty, but allows you to pass the transform even after the dataset is instantiated and then serialized. TBH, I feel like it would be better if we push to get batch transforms done in both vision and audio, bc I feel like if we had this, saving DS as an object wouldn't be (such) a problem.

File: references/video_classification/train.py (hunk 5 above)
Comment (bjuncek) [reply references/video_classification/train.py]: That's a neat trick - is there a benefit to doing this as opposed to the utility that you use in `maskrcnn_benchmark`?

Comment (fmassa) [reply references/video_classification/train.py]: This is just much simpler, and makes less assumptions, so I'm preferring this now over the implementation I had in `maskrcnn-benchmark`

File: references/video_classification/train.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        args.start_epoch = checkpoint['epoch'] + 1
+
+    if args.test_only:
+        evaluate(model, criterion, data_loader_test, device=device)
+        return
+
+    print("Start training")
+    start_time = time.time()
+    for epoch in range(args.start_epoch, args.epochs):
+        if args.distributed:
```
Comment (bjuncek) [inline references/video_classification/train.py]: Why don't you set validation sampler's epoch as well?

Comment (fmassa) [reply references/video_classification/train.py]: because there is no randomness for evaluation.

the `set_epoch` just makes the order of the examples to be different for each epoch, so there is no need to do it for validation.

File: torchvision/io/video.py (hunk 7 above)
Comment (fmassa) [reply torchvision/io/video.py]: I'm adding a TODO and removing the print for now

File: torchvision/datasets/kinetics.py [hunk 10]
```diff
@@ -23,4 +24,7 @@ def __getitem__(self, idx):
         video, audio, info, video_idx = self.video_clips.get_clip(idx)
         label = self.samples[video_idx][1]
 
+        if self.transform is not None:
```
Comment (fmassa) [inline torchvision/datasets/kinetics.py]: @bjuncek @soumith Looking for your opinions on how we should call the video transforms.

Indeed, we might have both `video_transform` and `audio_transform`, given that the dataset returns both data.

Would you prefer to stick with `transform` for meaning `video_transform`, or choose a different name?

File: references/video_classification/train.py (hunk 4 above)
Comment (fmassa) [reply references/video_classification/train.py]: I'll move `RandomClipSampler` in a follow-up PR

File: torchvision/datasets/kinetics.py (hunk 6 above)
Comment (fmassa) [reply torchvision/datasets/kinetics.py]: Improving the validation to not take into account duplicate indices will be done in a follow-up PR.

File: references/video_classification/train.py (hunk 3 above)
Comment (fmassa) [reply references/video_classification/train.py]: I added some basic caching mechanism, which can be improved after the release.

File: torchvision/datasets/kinetics.py (hunk 10 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py]: For this PR, I think we should keep this as transform, and maybe have a wrapper for audio transforms, or wait until batched transforms for both audio and video are ready

**Comment by fmassa** [comment]
For reference and to complement @bjuncek note, training on Kinetics400 for `r2plus1d_18` gives the following results for `clip-len=16`:

```
Clip Acc@1 57.351 Clip Acc@5 78.523
```

which matches the expected results from https://github.com/facebookresearch/VMZ/blob/master/tutorials/models.md
